include/cpp-statsd-client/StatsdClient.hpp

```diff
@@ -67,25 +70,25 @@ class StatsdClient {
     std::string m_prefix;
 
     //! The UDP sender to be used for actual sending
-    mutable UDPSender m_sender;
+    std::unique_ptr<UDPSender> m_sender;
 };
 
 inline StatsdClient::StatsdClient(const std::string& host,
                                   const uint16_t port,
                                   const std::string& prefix,
-                                  const std::optional<uint64_t> batchsize) noexcept
-    : m_prefix(prefix), m_sender(host, port, batchsize) {
-    // Initialize the randorm generator to be used for sampling
+                                  const uint64_t batchsize) noexcept
+    : m_prefix(prefix), m_sender(new UDPSender{host, port, batchsize}) {
+    // Initialize the random generator to be used for sampling
     std::srand(time(nullptr));
 }
 
-inline void StatsdClient::setConfig(const std::string& host, const uint16_t port, const std::string& prefix) noexcept {
+inline void StatsdClient::setConfig(const std::string& host, const uint16_t port, const std::string& prefix, const uint64_t batchsize) noexcept {
     m_prefix = prefix;
-    m_sender.setConfig(host, port);
+    m_sender.reset(new UDPSender(host, port, batchsize));
```

Comment:
so i changed this bit of the underlying implementation an i'm not sure what you'll think of it so please let me know.

initialization of the socket was done lazily before. basicaly when you wanted to send something it would first check if you've initialized, until that point it wouldnt bother with initializing. this causes a couple of problems:

1. when testing you basically cant tell that anything happened when you construct the object, initializing right away allows you to fail fast and detect it (and error or handle the error)
2. there is a race between `sendToDaemon` and `setConfig` the sender. the latter and tear down the socket while the former is in the middle of sending one of the messages batching mode (ie from the batching thread). we could have fixed this with synchronization but it seemed easier to just initialize once and let that be the config for the life time of the object. the top level interface still allows you to "re-configure" but under the hood it just makes a new `UDPSender`

Reply:
Hmm I can't remember why I went for lazy initialization back then :thinking: 
I'm fine changing this since it makes the code simpler and easier to test.

Using the `unique_ptr` and `reset` is the right thing to do, thanks :+1: